File: src/ObjectManager/Object.Tes/FilePacks/BsaMultiFile+Extensions.cs
using OA.Core;
using OA.Formats;
using OA.Tes.Formats;
using System;
using System.IO;
using System.Threading.Tasks;

namespace OA.Tes.FilePacks
{
    partial class BsaMultiFile
    {
        public Task<Texture2DInfo> LoadTextureInfoAsync(string texturePath)
        {
            var filePath = FindTexture(texturePath);
            if (filePath != null)
            {
                var fileData = LoadFileData(filePath);
                return Task.Run(() =>
                {
                    var fileExtension = Path.GetExtension(filePath);
                    if (fileExtension.ToLowerInvariant() == ".dds") return DdsReader.LoadDDSTexture(new MemoryStream(fileData));
                    else throw new NotSupportedException($"Unsupported texture type: {fileExtension}");
                });
            }
            else
            {
                Utils.Warning($"Could not find file \"{texturePath}\" in a BSA file.");
                return Task.FromResult<Texture2DInfo>(null);
            }
        }

        public Task<object> LoadObjectInfoAsync(string filePath)
        {
            var fileData = LoadFileData(filePath);
            return Task.Run(() =>
            {
                var file = new NiFile(Path.GetFileNameWithoutExtension(filePath));
                file.Deserialize(new UnityBinaryReader(new MemoryStream(fileData)));
                return (object)file;
            });
        }

        /// <summary>
        /// Finds the actual path of a texture.
        /// </summary>
        string FindTexture(string texturePath)
        {
            var textureName = Path.GetFileNameWithoutExtension(texturePath);
            var textureNameInTexturesDir = "textures/" + textureName;
            var filePath = textureNameInTexturesDir + ".dds";
            if (ContainsFile(filePath))
                return filePath;
            //filePath = textureNameInTexturesDir + ".tga";
            //if (ContainsFile(filePath))
            //    return filePath;
            var texturePathWithoutExtension = Path.GetDirectoryName(texturePath) + '/' + textureName;
            filePath = texturePathWithoutExtension + ".dds";
            if (ContainsFile(filePath))
                return filePath;
            //filePath = texturePathWithoutExtension + ".tga";
            //if (ContainsFile(filePath))
            //    return filePath;
            // Could not find the file.
            return null;
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/BsaMultiFile+Extensions.swift
//
//  BsaMultiFile+Extensions.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation

extension BsaMultiFile {
    public func loadTextureInfoAsync(texturePath: String) -> Task<Texture2DInfo?> {
        guard let filePath = findTexture(texturePath) else {
            debugPrint("Could not find file '\(texturePath)' in a BSA file.")
            return Task(value: nil)
        }
        let fileData = loadFileData(filePath)
        var task = Task<Texture2DInfo?>()
        DispatchQueue.global().async {
            debugPrint("\(filePath)")
            let fileExtension = URL(string: filePath)!.pathExtension
            guard fileExtension.lowercased() == "dds" else {
                fatalError("Unsupported texture type: \(fileExtension)")
            }
            let r = BinaryReader(DataBaseStream(data: fileData))
            defer { r.close() }
            let texture = DdsReader.loadDDSTexture(r)
            task.callback(texture)
        }
        return task
    }

    public func loadObjectInfoAsync(filePath: String) -> Task<Any> {
        
        let fileData = loadFileData(filePath)
        var task = Task<Any>()
        DispatchQueue.global().async {
            debugPrint("\(filePath)")
            let fileUrl = URL(string: filePath.replacingOccurrences(of: "\\", with: "/"))!
            let r = BinaryReader(DataBaseStream(data: fileData))
            defer { r.close() }
            let file = NiFile(r, name: fileUrl.deletingPathExtension().lastPathComponent)
            task.callback(file)
        }
        return task
    }
    
    func findTexture(_ texturePath: String) -> String? {
        let textureUrl = URL(string: texturePath)!
        let textureName = textureUrl.deletingPathExtension().lastPathComponent
        let textureNameInTexturesDir = "textures/\(textureName)"
        var filePath = "\(textureNameInTexturesDir).dds"
        if containsFile(filePath) {
            return filePath
        }
        let texturePathWithoutExtension = "\(textureUrl.deletingLastPathComponent())\\\(textureName)"
        filePath = "\(texturePathWithoutExtension).dds"
        if containsFile(filePath) {
            return filePath
        }
        return nil
    }
}
